pybind_interface/pybind_main.cpp

```diff
@@ -448,297 +448,199 @@ std::vector<std::complex<float>> qtrajectory_simulate(const py::dict &options) {
+    params.verbosity = verbosity;
+    return params;
   }
-  return state;
-}
 
-qsim::Simulator<For>::State _noisy_sim_from_input_state(
-    const py::dict &options, uint64_t input_state) {
-  using Simulator = qsim::Simulator<For>;
-  using StateSpace = Simulator::StateSpace;
-  using State = StateSpace::State;
-
-  NoisyCircuit<Cirq::GateCirq<float>> ncircuit;
-  unsigned num_qubits;
-  try {
-    ncircuit = getNoisyCircuit(options);
-    num_qubits = parseOptions<unsigned>(options, "n\0");
-  } catch (const std::invalid_argument &exp) {
-    IO::errorf(exp.what());
-    return StateSpace::Null();
-  }
-
-  using Runner = qsim::QuantumTrajectorySimulator<IO, Cirq::GateCirq<float>,
-                                                  MultiQubitGateFuser,
-                                                  Simulator>;
-  Runner::Parameter param;
-  uint64_t seed;
-  try {
-    param.num_threads = parseOptions<unsigned>(options, "t\0");
-    param.max_fused_size = parseOptions<unsigned>(options, "f\0");
-    param.verbosity = parseOptions<unsigned>(options, "v\0");
-    seed = parseOptions<unsigned>(options, "s\0");
-  } catch (const std::invalid_argument &exp) {
-    IO::errorf(exp.what());
-    return StateSpace::Null();
-  }
-
-  StateSpace state_space(param.num_threads);
-  State state = state_space.Create(num_qubits);
-  state_space.SetAllZeros(state);
-  state_space.SetAmpl(state, input_state, 1, 0);
-
-  State scratch = StateSpace(1).Null();
-  std::vector<uint64_t> stat;
-  if (!Runner::Run(param, num_qubits, ncircuit, seed, scratch, state, stat)) {
-    IO::errorf(
-      "qtrajectory full state simulation of the circuit errored out.\n");
-    return StateSpace::Null();
-  }
-  return state;
-}
-
-qsim::Simulator<For>::State _noisy_sim_from_input_vector(
-    const py::dict &options, const py::array_t<float> &input_vector) {
-  using Simulator = qsim::Simulator<For>;
-  using StateSpace = Simulator::StateSpace;
-  using State = StateSpace::State;
-
-  NoisyCircuit<Cirq::GateCirq<float>> ncircuit;
-  unsigned num_qubits;
-  try {
-    ncircuit = getNoisyCircuit(options);
-    num_qubits = parseOptions<unsigned>(options, "n\0");
-  } catch (const std::invalid_argument &exp) {
-    IO::errorf(exp.what());
-    return StateSpace::Null();
-  }
-
-  using Runner = qsim::QuantumTrajectorySimulator<IO, Cirq::GateCirq<float>,
-                                                  MultiQubitGateFuser,
-                                                  Simulator>;
-  Runner::Parameter param;
-  uint64_t seed;
-  try {
-    param.num_threads = parseOptions<unsigned>(options, "t\0");
-    param.max_fused_size = parseOptions<unsigned>(options, "f\0");
-    param.verbosity = parseOptions<unsigned>(options, "v\0");
-    seed = parseOptions<unsigned>(options, "s\0");
-  } catch (const std::invalid_argument &exp) {
-    IO::errorf(exp.what());
-    return StateSpace::Null();
+  template <typename StateType>
+  bool simulate(StateType input_state) {
```

Comment:
Should that be `const StateType& input_state`?

Reply:
Done, here and elsewhere in this file.